repo/subcmds/__init__.py

```diff
@@ -14,15 +14,100 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import os
+import sys
 import pkg_resources
+import pkgutil
+import sys
+
+from typing import Generator
+from typing import Set
+
+SubmoduleName = str
+
+def _get_module_by_name(namespace: str):
+  return sys.modules[namespace]
+
+
+def iter_namespace(namespace: SubmoduleName) -> Generator[SubmoduleName, None, None]:
+  """Pyinstaller-compatible namespace iteration.
+
+  Yields the name of all modules found at a given full-qualified module name.
+  Use 'importlib.import_module' to import the found submodules.
+
+  Usage:
+
+  .. code-block:: python
+
+      import importlib
+      from alton_utils.pyinstaller import iter_namespace
+
+      for found_submod in iter_namespace("module.submodule_to_iter"):
+          m = importlib.import_module(found_submod)
+
+  To have it running with pyinstaller, it still requires to ensure the wanted library
+  is well packaged in your executable, and since there is no proper 'import' statements
+  in your code pyinstaller might not find them. You might want to add a hook inject the
+  "hidden" modules from your plugins folder inside the executable:
+
+  - if your plugins are under the ``myappname/pluginfolder`` module
+  - create a file ``specs/hook-<myappname.pluginfolder>.py``
+  - content of this file should be:
+
+      .. code-block:: python
+
+          from PyInstaller.utils.hooks import collect_submodules
+          hiddenimports = collect_submodules('<myappname.pluginfolder>')
+  """
+
+  ns_pkg = _get_module_by_name(namespace)
+
+  prefix = ns_pkg.__name__ + "."
+  for p in pkgutil.iter_modules(ns_pkg.__path__, prefix):
+    yield p[1]
+
+
+  # special handling when the package is bundled with PyInstaller 3.5
+  # See https://github.com/pyinstaller/pyinstaller/issues/1905#issuecomment-445787510
+  toc: Set[str] = set()
+  for importer in pkgutil.iter_importers(ns_pkg.__name__.partition(".")[0]):
+    name = getattr(importer, "toc", None)
+    if name:
+      toc |= name
+  for name in toc:
+    if not name.startswith(prefix):
+      continue
+    rem_name = name[len(prefix) :]
+    if "." in rem_name:
+      continue
+    yield name
 
 all_commands = {}
-for ep in pkg_resources.iter_entry_points("repo.subcmds"):
-  cmd = ep.resolve()
-  cmdname = ep.name
-  cmd.NAME = cmdname
-  all_commands[cmdname] = cmd
+eps = list(pkg_resources.iter_entry_points("repo.subcmds"))
+if eps:
+  for ep in eps:
+    cmd = ep.resolve()
+    cmdname = ep.name
+    cmd.NAME = cmdname
+    all_commands[cmdname] = cmd
+else:
+  for modname in iter_namespace(__name__):
+    name = modname.rpartition('.')[2]
+    clsn = name.capitalize()
+    while clsn.find('_') > 0:
+      h = clsn.index('_')
+      clsn = clsn[0:h] + clsn[h + 1:].capitalize()
+
+    import importlib
+    mod = importlib.import_module(modname)
```

Comment:
the whole point of iter_entry_points is to not load the modules (and their dependencies).

Then you are building here a heuristic to map classnames to subcommand names which looks like very fragile.

I doubt I will be able to upstream this when the time comes.

Reply:
1/ I just added a fallback when iter_entry_points() has no return
2/ This fallback code is based on [original code](https://github.com/tardyp/google-repo/commit/ea11d8c6d44960b6b01455d089bfe52b3b423509#diff-5db94cc59acc7877994e1dcf547b9899155e6e1f536f8c91b7ccf22c823eb29aL33), would you prefere I use __import__() ?

Reply:
In this case, this indeed makes sense. I think this deserve a comment.